fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java

```diff
@@ -39,47 +46,130 @@ public StarOSAgent() {
+        for (ReplicaInfo replicaInfo : replicaInfos) {
+            WorkerInfo workerInfo = replicaInfo.getWorkerInfo();
+            String ipPort = workerInfo.getIpPort();
+            String host = ipPort.split(":")[0];
+            long workerId = workerToId.get(ipPort);
+            // String host = ipPort.split(":")[0];
+            // long backendId = GlobalStateMgr.getCurrentSystemInfo().getBackendIdByHost(host);
+            if (!workerToBackend.containsKey(workerId)) {
+                LOG.warn("Backend does not exists. host: {}", host);
+                continue;
+            }
+            long backendId = workerToBackend.get(workerId);
             backendIds.add(backendId);
         }
         return backendIds;
     }
 
     public void registerAndBootstrapService(String serviceName) {
-        if (serviceId == -1) {
+        if (serviceId != -1) {
+            return;
+        }
+
+        try {
             client.registerService("starrocks");
+        } catch (StarClientException e) {
+            if (e.getCode() != StarClientException.ExceptionCode.ALREADY_EXIST) {
+                LOG.warn(e);
+                System.exit(-1);
+            }
+        }
+
+        try {
             serviceId = client.bootstrapService("starrocks", serviceName);
+            LOG.info("get serviceId: {} by bootstrapService to strMgr", serviceId);
+        } catch (StarClientException e) {
+            if (e.getCode() != StarClientException.ExceptionCode.ALREADY_EXIST) {
+                LOG.warn(e);
+                System.exit(-1);
+            } else {
+                getServiceId(serviceName);
+            }
         }
-        LOG.info("serviceId from starClient is {} ", serviceId);
     }
 
     public void getServiceId(String serviceName) {
         if (serviceId != -1) {
             return;
         }
-        serviceId = client.getServiceInfo(serviceName);
+        try {
+            ServiceInfo serviceInfo = client.getServiceInfo(serviceName);
+            serviceId = serviceInfo.getServiceId();
+        } catch (StarClientException e) {
+            LOG.warn(e);
+            System.exit(-1);
+        }
+        LOG.info("get serviceId: {} by getServiceInfo from strMgr", serviceId);
     }
 
+    public void addWorker(long backendId, String workerIpPort) {
+        if (serviceId == -1) {
+            LOG.warn("When addWorker serviceId is -1");
+            return;
+        }
+        if (workerToId.containsKey(workerIpPort)) {
+            LOG.info("workerToId contains {}, workerId is {}", workerIpPort, workerToId.get(workerIpPort));
+            return;
+        }
+        long workerId = -1;
+        try {
+            workerId = client.addWorker(serviceId, workerIpPort);
+        } catch (StarClientException e) {
+            if (e.getCode() != StarClientException.ExceptionCode.ALREADY_EXIST) {
+                LOG.warn(e);
+            } else {
+                workerToId.put(workerIpPort, workerId);
```

Comment:
workerId is invalid when catching exception.
Maybe we should add a new api to get worker info by ip port.